### PR TITLE
osd/PrimaryLogPG: refresh local last-complete iter for async_recovery_targets properly

### PR DESCRIPTION
**Pull request by xiexingguo**
Upon overwritting an already-missing object on async_recovery_targets,
besides updating the **missing** set async_recovery_targets should
call reset_complete_to() simultaneously to reset **complete-to** iter and
advance **last_complete** properly, as those old log entries which we used
to rely on for recovery are now getting replaced by new **committed** ones.

Since the recovery process of an async_recovery_target can be arbitrarily
delayed until we hit the osd_max_pg_log_entries limit, the main befenit
of an extra call to reset_complete_to() here is to make sure primary perform
log-trim in time. Otherwise accumulating of huge amounts of **obsolete** log
entries can slow the system down obviously.

Signed-off-by: xiexingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

**Comment by xiexingguo** [comment]
@jdurgin

**Comment by xiexingguo** [comment]
@tchaikov 

This change is currently triggering crashes.  Please drop it from your test branch. :-(

**Comment by xiexingguo** [comment]
@neha-ojha 

I appended another fix to address the crashes from @tchaikov 's test branch and the newest test results seem to be positive:

http://pulpito.ceph.com/xxg-2018-04-25_06:14:51-rados:thrash-wip-refresh-lc-distro-basic-smithi/

http://pulpito.ceph.com/xxg-2018-04-25_04:38:44-rados-wip-refresh-lc-distro-basic-smithi/

Mind taking another glance at?

File: src/osd/ECBackend.cc [hunk 1]
```diff
@@ -902,7 +902,16 @@ void ECBackend::handle_sub_write(
     }
   }
   clear_temp_objs(op.temp_removed);
-  dout(30) << __func__ << " missing before " << get_parent()->get_log().get_missing().get_items() << dendl;
+  get_parent()->log_operation(
```
Comment (neha-ojha) [inline src/osd/ECBackend.cc]: Is there is reason why we want to move the dout of "missing before"?

Comment (xiexingguo) [reply src/osd/ECBackend.cc]: dropped that cosmetic change ( I was trying to wrap that line into 80 chars, but I think it is fine to leave as it is...)

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-04-30_00:59:17-rados-wip-kefu-testing-2018-04-29-1248-distro-basic-smithi/2454381/

could be relevant .

**Comment by liewegas** [comment]
This looks like itw as causing scrub errors in my run

**Pull request closed without merging**